### PR TITLE
Move a bunch of Linalg docs out of helpdb, improve math docs

### PR DESCRIPTION
**Pull request by kshyatt**
Made docs for `bkfact`, `hessenberg`, `lu` more accurate.
Added examples for more functions.
Moved some special functions out of helpdb as well.

### CODE REVIEW

File: base/linalg/bunchkaufman.jl [hunk 1]
```diff
@@ -16,6 +16,12 @@ BunchKaufman{T}(A::AbstractMatrix{T}, ipiv::Vector{BlasInt}, uplo::Char, symmetr
     rook::Bool, info::BlasInt) =
         BunchKaufman{T,typeof(A)}(A, ipiv, uplo, symmetric, rook, info)
 
+"""
+    bkfact!(A, uplo::Symbol=:U, symmetric::Bool=issymmetric(A), rook::Bool=false) -> BunchKaufman
+
+`bkfact!` is the same as [`bkfact`](:func:`bkfact`), but saves space by overwriting the
+input `A`, instead of creating a copy. `uplo`
```
Comment (kshyatt) [inline base/linalg/bunchkaufman.jl]: Ooops.

File: base/docs/helpdb/Base.jl [hunk 2]
```diff
@@ -1497,10 +1574,17 @@ selectperm
...
-`Float32`. For arrays, this constructs an array with the same binary data as the given
+Change the type-interpretation of a block of memory.
+For arrays, this constructs an array with the same binary data as the given
 array, but with the specified element type.
+For example,
+`reinterpret(Float32, UInt32(7))` interprets the 4 bytes corresponding to `UInt32(7)` as a
+`Float32`.
+
+```jldoctest
+julia> reinterpret(Float32, UInt32(7))
```
Comment (andreasnoack) [inline base/docs/helpdb/Base.jl]: Maybe also show an example with an array

File: base/docs/helpdb/Base.jl [hunk 3]
```diff
@@ -2288,6 +2380,14 @@ isreal
 
 Return `true` if and only if all values of `type1` are also of `type2`. Can also be written
 using the `<:` infix operator as `type1 <: type2`.
+
+```jldoctest
+julia> issubtype(Int8, Int32)
+false
+
+julia> issubtype(Int8, Integer)
```
Comment (andreasnoack) [inline base/docs/helpdb/Base.jl]: Maybe use the `<:` version. It's much more common.

File: base/docs/helpdb/Base.jl [hunk 4]
```diff
@@ -2350,6 +2450,33 @@ countlines
     .\\(x, y)
 
 Element-wise left division operator.
+
+```jldoctest
+julia> A = [1 0; 0 -1];
```
Comment (andreasnoack) [inline base/docs/helpdb/Base.jl]: Seems like an advanced/complicated example. Most often, this is probably used for scalar elements.

Comment (kshyatt) [inline base/docs/helpdb/Base.jl]: @tkelman didn't like my simple examples for `.+` and `.*` so I'm going all out. I think this is a good example to show people some cool things Julia can do. If they get too confused they can go to `broadcast` and learn so many wonderful things.

File: base/linalg/bunchkaufman.jl [hunk 5]
```diff
@@ -47,6 +53,22 @@ function bkfact!{T<:BlasComplex}(A::StridedMatrix{T}, uplo::Symbol=:U,
     end
     BunchKaufman(LD, ipiv, char_uplo(uplo), symmetric, rook, info)
 end
+
+"""
+    bkfact(A, uplo::Symbol=:U, symmetric::Bool=issymmetric(A), rook::Bool=false) -> BunchKaufman
+
+Compute the Bunch-Kaufman [^Bunch1977] factorization of a real symmetric or complex Hermitian
```
Comment (andreasnoack) [inline base/linalg/bunchkaufman.jl]: Can also be complex symmetric.

File: base/docs/helpdb/Base.jl (hunk 4 above)
Comment (andreasnoack) [inline base/docs/helpdb/Base.jl]: Can we have both?

File: base/docs/helpdb/Base.jl [hunk 6]
```diff
@@ -69,7 +81,15 @@ Base.:(//)
...
     isinteger(x) -> Bool
 
-Test whether `x` or all its elements are numerically equal to some integer
+Test whether `x` or all its elements are numerically equal to some integer.
+
+```jldoctest
+julia> isinteger(4.0)
+true
+
+julia> isinteger([1; 2; 5.5])
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: we should think about deprecating the vectorized version of this at some point

File: base/docs/helpdb/Base.jl [hunk 7]
```diff
@@ -426,6 +477,14 @@ filter!
     sizeof(T)
 
 Size, in bytes, of the canonical binary representation of the given DataType `T`, if any.
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: what happens if not any?

File: base/docs/helpdb/Base.jl (hunk 4 above)
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: elementwise left division needs an interesting example, it's not so useful for simple cases

File: base/docs/helpdb/Base.jl [hunk 8]
```diff
@@ -2518,21 +2653,19 @@ Base.:(!)
...
+```jldoctest
+julia> length(1:5)
+5
 
+julia> length([1; 2; 3; 4])
+4
+```
 """
+length(collection)
 bkfact
```
Comment (tkelman) [inline base/docs/helpdb/Base.jl]: this `bkfact` shouldn't be here any more

File: base/linalg/lu.jl [hunk 9]
```diff
@@ -132,6 +141,13 @@ lufact(x::Number) = LU(fill(x, 1, 1), BlasInt[1], x == 0 ? one(BlasInt) : zero(B
 lufact(F::LU) = F
 
 lu(x::Number) = (one(x), x, 1)
+
+"""
+    lu(A) -> L, U, p
```
Comment (tkelman) [inline base/linalg/lu.jl]: should `pivot` be documented or is that only intended for going through `lufact` going forward?

Comment (andreasnoack) [inline base/linalg/lu.jl]: I think it should be documented for `lu` as well.

**Comment by kshyatt** [comment]
Good to go if lights turn green?

File: base/linalg/bunchkaufman.jl (hunk 5 above)
Comment (andreasnoack) [inline base/linalg/bunchkaufman.jl]: This is wrong so it should be fixed at some point. Might make sense to do after you have moved the doc string though.

**Comment by kshyatt** [comment]
I think I addressed all the comments, CI passed, it's been a day.

**Comment by andreasnoack** [comment]
What are your thoughts on fixing the `bkfact` doc string?

**Comment by kshyatt** [comment]
@andreasnoack I'll do it in the other linalg PR I have open?

**Pull request merged**